src/coins.cpp

```diff
@@ -186,30 +186,34 @@ bool CCoinsViewCache::BatchWrite(CoinsViewCacheCursor& cursor, const uint256 &ha
         if (!it->second.IsDirty()) {
             continue;
         }
-        CCoinsMap::iterator itUs = cacheCoins.find(it->first);
-        if (itUs == cacheCoins.end()) {
+        CCoinsMap::iterator itUs;
+        bool isInserted = false;
+        if (!(it->second.IsFresh() && it->second.coin.IsSpent())) {
+            std::tie(itUs, isInserted) = cacheCoins.try_emplace(it->first);
+        } else {
+            itUs = cacheCoins.find(it->first);
```

Comment:
It looks like this branch isn't needed, in which case we can always just `try_emplace` (and add a dead-code erase for the tests), see [`ae76ec7` (#30673)](https://github.com/bitcoin/bitcoin/pull/30673/commits/ae76ec7bcff0a08a61f294882a71e46d177b009f#diff-f0ed73d62dae6ca28ebd3045e5fc0d5d02eaaacadb4c2a292985a3fbd7e1c77cR184)

I've pushed it to my own fork for now (https://github.com/l0rinc/bitcoin/pull/34), adding you and @andrewtoth as coauthors. Will publish once my benchmarks finish.